### PR TITLE
Demonstrate usage of WAM (and update to MSAL 4.23)

### PR DESCRIPTION
**Pull request by jmprieur**
Update to MSAL.NET 4.20 and showcase the usage of WAM.
- Adds a combo box to use WAM (default account, any WAM account, any Azure AD Account)
- Updates the app creation depending on the combo box's selected option
- and update `CallGraphButton_Click` to enable any WAM account

### CODE REVIEW

File: active-directory-wpf-msgraph-v2/MainWindow.xaml.cs [hunk 1]
```diff
@@ -37,9 +42,20 @@ private async void CallGraphButton_Click(object sender, RoutedEventArgs e)
...
+            // WAM will always get an account in the cache. So if we want
+            // to have a chance to select the accounts interactively, we need to
+            // force the non-account
+            if (howToSignIn.SelectedIndex == 1)
+            {
+                firstAccount = null;
+            }
+            else
+            {
+                var accounts = await app.GetAccountsAsync();
```
Comment (bgavrilMS) [inline active-directory-wpf-msgraph-v2/MainWindow.xaml.cs]: This (`firstOrDefault`) is a bit artificial. In my test apps, I have a drop-down control with all known accounts and users can either select a "known" account or they select none. I think this is closer to real-world scenario than "randomly pick an account". I can contribute this if you wish to go down this path.

File: active-directory-wpf-msgraph-v2/MainWindow.xaml.cs [hunk 2]
```diff
@@ -37,9 +42,20 @@ private async void CallGraphButton_Click(object sender, RoutedEventArgs e)
...
+                firstAccount = accounts.FirstOrDefault();
+            }
```
Comment (bgavrilMS) [inline active-directory-wpf-msgraph-v2/MainWindow.xaml.cs:63]: We should have a discussion about prompts as well. In WAM, there are only 2 prompts - "default" and "force". "Force" is very similar to our "select_account". 

The current logic I implemented for mapping is as follows: 

https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/882c815c5e319b5ab356cca1bbfc39ac36dd022b/src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/WamBroker.cs#L143


- if the user configures `select_account` or `force_login`, wam prompt becomes `force`
- if the user configures `no_prompt` or does not make any configuration, WAM prompt becomes `default`
- all other prompts cannot be used

Happy to re-discuss this point. I think MSAL should also stop defaulting to "select_account" and instead it should not configure any prompt.

File: active-directory-wpf-msgraph-v2/MainWindow.xaml.cs [hunk 3]
```diff
@@ -37,9 +42,20 @@ private async void CallGraphButton_Click(object sender, RoutedEventArgs e)
...
+            IAccount firstAccount;
 
+            // WAM will always get an account in the cache. So if we want
+            // to have a chance to select the accounts interactively, we need to
+            // force the non-account
+            if (howToSignIn.SelectedIndex == 1)
+            {
+                firstAccount = null;
+            }
+            else
```
Comment (bgavrilMS) [inline active-directory-wpf-msgraph-v2/MainWindow.xaml.cs]: if (howToSingIn.SelectedIndex!=1) { ... }

**Review by bgavrilMS — APPROVED** [review]

**Comment by jmprieur** [comment]
@bgavrilMS should we postpone updating this PR?

**Comment by bgavrilMS** [comment]
@jmprieur - yeah, let's postpone

File: active-directory-wpf-msgraph-v2/MainWindow.xaml [hunk 4]
```diff
@@ -5,10 +5,15 @@
...
         mc:Ignorable="d"
-        Title="MainWindow" Height="Auto" Width="525" SizeToContent="Height">
+        Title="MainWindow" Height="Auto" Width="600" SizeToContent="Height">
     <Grid>
         <StackPanel Background="Azure">
             <StackPanel Orientation="Horizontal" HorizontalAlignment="Right">
+                <ComboBox x:Name="howToSignIn" SelectedIndex="0" SelectionChanged="UseWam_Changed" VerticalAlignment="Center">
+                    <ComboBoxItem Tag="0">Use account used to signed-in in Windows (WAM)</ComboBoxItem>
+                    <ComboBoxItem Tag="1">Use one of the Accounts known by Windows (WAM)</ComboBoxItem>
+                    <ComboBoxItem Tag="1">Use any account (Azure AD)</ComboBoxItem>
```
Comment (henrik-me) [inline active-directory-wpf-msgraph-v2/MainWindow.xaml]: >1 [](start = 39, length = 1)

should this be 2?

File: active-directory-wpf-msgraph-v2/MainWindow.xaml.cs [hunk 5]
```diff
@@ -27,6 +27,11 @@ public MainWindow()
             InitializeComponent();
         }
 
+        private void UseWam_Changed(object sender, RoutedEventArgs e)
+        {
+            App.CreateApplication(howToSignIn.SelectedIndex != 2);
```
Comment (henrik-me) [inline active-directory-wpf-msgraph-v2/MainWindow.xaml.cs]: >howToSignIn.SelectedIndex != 2 [](start = 34, length = 30)

should this be comparing to a tag value instead making the code more readable?

File: active-directory-wpf-msgraph-v2/MainWindow.xaml.cs (hunk 2 above)
Comment (jmprieur) [reply active-directory-wpf-msgraph-v2/MainWindow.xaml.cs:63]: @bgavrilMS : did you have an opportunity to discuss `Account.WindowsSession` with the WAM people? (or `IPublicClientApplication.WindowsSession`)

File: active-directory-wpf-msgraph-v2/MainWindow.xaml (hunk 4 above)
Comment (jmprieur) [reply active-directory-wpf-msgraph-v2/MainWindow.xaml]: yes, absolutely. Thanks @henrik-me

File: active-directory-wpf-msgraph-v2/MainWindow.xaml.cs (hunk 5 above)
Comment (jmprieur) [reply active-directory-wpf-msgraph-v2/MainWindow.xaml.cs]: Actually the tag as I wrote it didn't work (it needs a nested element, and expressing string). I removed the tags from the XAML file

File: active-directory-wpf-msgraph-v2/MainWindow.xaml.cs (hunk 3 above)
Comment (jmprieur) [reply active-directory-wpf-msgraph-v2/MainWindow.xaml.cs]: not for the moment (until null uses the default accounts. This was not the case in 4.20 yet

**Comment by bgavrilMS** [comment]
@jmprieur @henrik-me - after discussing with both @jmprieur and Shane, I'm now convinced that we should use

`AcquireTokenSilent(PublicClientApplication.WamDefaultAccount, scopes)` instead of null. Please vote for it here: https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview/pullrequest/2322

**Comment by jmprieur** [comment]
@bgavrilMS : you might be interested in https://github.com/Azure-Samples/active-directory-dotnet-desktop-msgraph-v2/pull/60/commits/9a1b3e86683bc9c1aed337829c4831b610c9dd61

File: active-directory-wpf-msgraph-v2/MainWindow.xaml.cs [hunk 6]
```diff
@@ -47,7 +42,13 @@ private async void CallGraphButton_Click(object sender, RoutedEventArgs e)
...
             // to have a chance to select the accounts interactively, we need to
             // force the non-account
-            if (howToSignIn.SelectedIndex == 1) // Account signed-in with Windows
+            if (howToSignIn.SelectedIndex == 0) // Account signed-in with Windows
+            {
+                firstAccount = PublicClientApplication.OperatingSystemAccount;
+            }
+
+            // WAM
+            else if (howToSignIn.SelectedIndex == 1)
```
Comment (bgavrilMS) [inline active-directory-wpf-msgraph-v2/MainWindow.xaml.cs]: Please add a comment like the one at line 45, explaining what index 1 and 2 are (i.e. what label it corresponds to)

**Review by bgavrilMS — APPROVED** [review]

**Review by bgavrilMS — APPROVED** [review]

**Pull request merged**